File: experiments/robust_blink_v2/head_motion.py
"""
Head motion tracking from facial landmarks.

Tracks head position using nose tip and face corners to detect movement.
Used to gate blink detection during head motion (which causes false EAR changes).
"""

from dataclasses import dataclass
from typing import Optional, Tuple
import math

from constants import (
    NOSE_TIP_INDEX,
    LEFT_FACE_CORNER_INDEX,
    RIGHT_FACE_CORNER_INDEX,
    HEAD_MOTION_THRESHOLD,
    HEAD_POSITION_SMOOTHING,
)


@dataclass
class HeadMotionResult:
    """Result of head motion analysis."""
    velocity: float           # Normalized velocity (face-widths per frame)
    is_moving: bool           # True if velocity exceeds threshold
    position: Tuple[float, float]  # Current (x, y) position
    face_width: float         # Current face width (for normalization)


class HeadMotionTracker:
    """
    Tracks head motion by monitoring facial landmark positions.

    Uses nose tip as primary position indicator, normalized by face width
    to be scale-invariant (works at different distances from camera).
    """

    def __init__(self, smoothing_alpha: float = HEAD_POSITION_SMOOTHING):
        self.smoothing_alpha = smoothing_alpha
        self.prev_position: Optional[Tuple[float, float]] = None
        self.smoothed_position: Optional[Tuple[float, float]] = None
        self.prev_face_width: Optional[float] = None

    def update(self, landmarks: list) -> HeadMotionResult:
        """
        Update head position and calculate velocity.

        Args:
            landmarks: List of MediaPipe NormalizedLandmark objects

        Returns:
            HeadMotionResult with velocity and motion flag
        """
        # Extract key landmarks
        nose = landmarks[NOSE_TIP_INDEX]
        left_corner = landmarks[LEFT_FACE_CORNER_INDEX]
        right_corner = landmarks[RIGHT_FACE_CORNER_INDEX]

        # Current position (nose tip)
        current_pos = (nose.x, nose.y)

        # Face width for normalization (distance between face corners)
        face_width = math.sqrt(
            (right_corner.x - left_corner.x) ** 2 +
            (right_corner.y - left_corner.y) ** 2
        )

        # Ensure face width is reasonable (avoid division issues)
        if face_width < 0.01:
            face_width = 0.1

        # Apply smoothing to position
        if self.smoothed_position is None:
            self.smoothed_position = current_pos
        else:
            self.smoothed_position = (
                self.smoothing_alpha * current_pos[0] +
                (1 - self.smoothing_alpha) * self.smoothed_position[0],
                self.smoothing_alpha * current_pos[1] +
                (1 - self.smoothing_alpha) * self.smoothed_position[1],
            )

        # Calculate velocity
        velocity = 0.0
        if self.prev_position is not None and self.prev_face_width is not None:
            # Distance moved
            dx = self.smoothed_position[0] - self.prev_position[0]
            dy = self.smoothed_position[1] - self.prev_position[1]
            distance = math.sqrt(dx * dx + dy * dy)

            # Normalize by face width (scale-invariant)
            avg_face_width = (face_width + self.prev_face_width) / 2
            velocity = distance / avg_face_width

        # Update state for next frame
        self.prev_position = self.smoothed_position
        self.prev_face_width = face_width

        # Check if moving
        is_moving = velocity > HEAD_MOTION_THRESHOLD

        return HeadMotionResult(
            velocity=velocity,
            is_moving=is_moving,
            position=self.smoothed_position,
            face_width=face_width,
        )

    def reset(self):
        """Reset tracking state. Call when face is lost."""
        self.prev_position = None
        self.smoothed_position = None
        self.prev_face_width = None

File: lumina/packages/core/src/detection/head-motion.ts
/**
 * Head motion tracking from facial landmarks.
 *
 * Tracks head position using nose tip and face corners to detect movement.
 * Used to gate blink detection during head motion (which causes false EAR changes).
 */

import { NOSE_TIP_INDEX, HEAD_MOTION } from './constants';
import { Point2D } from './blink';

export interface HeadMotionResult {
  /** Normalized velocity (face-widths per frame) */
  velocity: number;
  /** True if velocity exceeds threshold */
  isMoving: boolean;
  /** Current (x, y) position */
  position: { x: number; y: number };
  /** Current face width (for normalization) */
  faceWidth: number;
}

/**
 * Tracks head motion by monitoring facial landmark positions.
 *
 * Uses nose tip as primary position indicator, normalized by face width
 * to be scale-invariant (works at different distances from camera).
 */
export class HeadMotionTracker {
  private smoothingAlpha: number;
  private prevPosition: { x: number; y: number } | null = null;
  private smoothedPosition: { x: number; y: number } | null = null;
  private prevFaceWidth: number | null = null;

  constructor(smoothingAlpha: number = HEAD_MOTION.SMOOTHING_ALPHA) {
    this.smoothingAlpha = smoothingAlpha;
  }

  /**
   * Update head position and calculate velocity.
   *
   * @param landmarks Array of 478 MediaPipe face landmarks
   * @returns HeadMotionResult with velocity and motion flag
   */
  update(landmarks: Point2D[]): HeadMotionResult {
    // Extract key landmarks
    const nose = landmarks[NOSE_TIP_INDEX];
    const leftCorner = landmarks[HEAD_MOTION.LEFT_FACE_CORNER_INDEX];
    const rightCorner = landmarks[HEAD_MOTION.RIGHT_FACE_CORNER_INDEX];

    // Current position (nose tip)
    const currentPos = { x: nose.x, y: nose.y };

    // Face width for normalization (distance between face corners)
    let faceWidth = Math.sqrt(
      Math.pow(rightCorner.x - leftCorner.x, 2) +
      Math.pow(rightCorner.y - leftCorner.y, 2)
    );

    // Ensure face width is reasonable (avoid division issues)
    if (faceWidth < 0.01) {
      faceWidth = 0.1;
    }

    // Apply smoothing to position
    if (this.smoothedPosition === null) {
      this.smoothedPosition = currentPos;
    } else {
      this.smoothedPosition = {
        x: this.smoothingAlpha * currentPos.x +
           (1 - this.smoothingAlpha) * this.smoothedPosition.x,
        y: this.smoothingAlpha * currentPos.y +
           (1 - this.smoothingAlpha) * this.smoothedPosition.y,
      };
    }

    // Calculate velocity
    let velocity = 0;
    if (this.prevPosition !== null && this.prevFaceWidth !== null) {
      // Distance moved
      const dx = this.smoothedPosition.x - this.prevPosition.x;
      const dy = this.smoothedPosition.y - this.prevPosition.y;
      const distance = Math.sqrt(dx * dx + dy * dy);

      // Normalize by face width (scale-invariant)
      const avgFaceWidth = (faceWidth + this.prevFaceWidth) / 2;
      velocity = distance / avgFaceWidth;
    }

    // Update state for next frame
    this.prevPosition = { ...this.smoothedPosition };
    this.prevFaceWidth = faceWidth;

    // Check if moving
    const isMoving = velocity > HEAD_MOTION.THRESHOLD;

    return {
      velocity,
      isMoving,
      position: this.smoothedPosition,
      faceWidth,
    };
  }

  /**
   * Reset tracking state. Call when face is lost.
   */
  reset(): void {
    this.prevPosition = null;
    this.smoothedPosition = null;
    this.prevFaceWidth = null;
  }
}
